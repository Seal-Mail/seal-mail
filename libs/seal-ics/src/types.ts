import { CalendarComponent } from 'ical';
import { z } from 'zod';

export enum Params {
  Language = 'LANGUAGE',
  Name = 'CN',
  SentBy = 'SENT_BY',
  Role = 'ROLE',
  ParticipantStatue = 'PARTSTAT',
  Rsvp = 'RSVP',
  CalendarUserType = 'CUTYPE',
  Member = 'MEMBER',
  DelegatedTo = 'DELEGATED-TO',
  DelegatedFrom = 'DELEGATED-FROM'
}

export interface ParsedOrganizer {
  name: string;
  language?: string;
  sentBy?: string;
  value: string;
}

export enum AttendeeStatus {
  NeedActions = 'NEEDS-ACTION',
  Accepted = 'ACCEPTED',
  Declined = 'DECLINED',
  Tentative = 'TENTATIVE',
  Delegated = 'DELEGATED'
}
export const AttendeeStatParser = z.nativeEnum(AttendeeStatus).default(AttendeeStatus.NeedActions);

export enum AttendeeRole {
  Required = 'REQ-PARTICIPANT',
  Optional = 'OPT-PARTICIPANT'
}

export enum EventStatus {
  Tentative = 'TENTATIVE',
  Confirmed = 'CONFIRMED',
  Canceled = 'CANCELLED'
}

export enum CalendarUserTypes {
  Individual = 'INDIVIDUAL',
  Group = 'GROUP',
  Resource = 'RESOURCE',
  Room = 'ROOM',
  Unknown = 'UNKNOWN'
}

export const CalendarUserTypesParser = z.nativeEnum(CalendarUserTypes).default(CalendarUserTypes.Unknown);

// Types of the METHOD field described in the iTIP specifications.
export enum CalendarMethodTypes {
  Publish = 'PUBLISH',
  Request = 'REQUEST',
  Reply = 'REPLY',
  Add = 'ADD',
  Cancel = 'CANCEL',
  Refresh = 'REFRESH',
  Counter = 'COUNTER',
  DeclineCounter = 'DECLINECOUNTER'
}

export interface ParsedAttendee {
  name: string;
  value: string;
  email: string;
  sentBy?: string;
  language?: string;
  role: string;
  status: AttendeeStatus;
  rsvp: boolean;
  calendarUserType: CalendarUserTypes;
  memberships?: string;
  delegatedTo?: string;
  delegatedFrom?: string;
}

export type ExtendedDate = Date & {
  dateOnly?: boolean;
  tz?: string | number | boolean;
};

export type ExtendedCalendarComponent = CalendarComponent & {
  start?: ExtendedDate;
  end?: ExtendedDate;
};
