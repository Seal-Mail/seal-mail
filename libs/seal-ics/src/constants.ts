import { CalendarMethodTypes } from './types';

export const ICAL_COMPANY = 'Skiff';
export const ICAL_PRODUCT = 'Skiff Calendar';
export const ICAL_LANGUAGE = 'EN';
const joinedMethods = Object.values(CalendarMethodTypes).join('|');
export const METHOD_REGEX_PATTERN = `METHOD:(${joinedMethods})`;
export const GMT_TZID_PATTERN = 'GMT[+-]\\d{4}';

export const CONFERENCE_KEY = 'X-SKIFF-MEET';

// Time constants
export const SECONDS_IN_MIN = 60;
export const MINUTES_IN_HOUR = 60;
export const DEFAULT_ALL_DAY_REMINDER_HOURS_BEFORE = 15;
export const DEFAULT_TIMED_REMINDER_MINS_BEFORE = 10;

export const HOUR_UNIT = 'hour';
export const DAY_UNIT = 'day';
export const DAYS_IN_WEEK = 'week';
