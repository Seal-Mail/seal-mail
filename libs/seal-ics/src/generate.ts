import ical from 'ical-generator';
import {
  ICalAttendeeStatus,
  ICalEventStatus,
  ICalCalendarMethod,
  ICalAttendeeRole,
  ICalEventTransparency,
  ICalAlarmType
} from 'ical-generator';
import { AttendeePermission, AttendeeStatus } from 'skiff-graphql';
import isURL from 'validator/lib/isURL';

import {
  ICAL_COMPANY,
  ICAL_LANGUAGE,
  ICAL_PRODUCT,
  CONFERENCE_KEY,
  DEFAULT_ALL_DAY_REMINDER_HOURS_BEFORE,
  MINUTES_IN_HOUR,
  SECONDS_IN_MIN,
  DEFAULT_TIMED_REMINDER_MINS_BEFORE
} from './constants';
import { RecurrenceRule } from './RecurrenceRule';

export enum EventAttendeeType {
  InternalAttendee,
  ExternalAttendee
}

export type GenerateAttendee = {
  attendeeStatus: AttendeeStatus;
  permission: AttendeePermission;
  optional: boolean;
  email: string;
  displayName?: string;
};

export interface GenerateEvent {
  title: string;
  description?: string;
  startDate: number;
  endDate: number;
  externalID: string;
  location?: string;
  isAllDay?: boolean;
  attendees: GenerateAttendee[];
  updatedAt: number;
  sequence: number;
  conference?: string;
  recurrenceRule?: RecurrenceRule;
  recurrenceID?: Date;
}

const transformStatus = (status: AttendeeStatus): ICalAttendeeStatus => {
  switch (status) {
    case AttendeeStatus.Pending:
      return ICalAttendeeStatus.NEEDSACTION;
    case AttendeeStatus.Maybe:
      return ICalAttendeeStatus.TENTATIVE;
    case AttendeeStatus.Yes:
      return ICalAttendeeStatus.ACCEPTED;
    case AttendeeStatus.No:
      return ICalAttendeeStatus.DECLINED;
  }
};

/**
 * return local tz using Intl.DateTimeFormat
 * we always need the local timezone, not matter what the users skiff calendar app is configured for,
 * because the native `Date` object is always created with the machine timezone
 *
 * this is supported for 95% of users, in case it's not, return undefined
 * This is also what `dayjs` uses in `dayjs.tz.guess`
 * https://caniuse.com/?search=Intl.DateTimeFormat().resolvedOptions().timeZone
 * @returns
 */
const getLocalTimezone = () => {
  try {
    const tz = Intl.DateTimeFormat().resolvedOptions().timeZone;
    return tz;
  } catch {
    return undefined;
  }
};

/**
 * Adds the conference parameter in the location parameter so it is visible for external users
 */
const constructLocationString = (location: string | undefined, conference: string | undefined) => {
  if (location && conference) {
    return `${location}, ${conference}`;
  } else if (conference) {
    return conference;
  } else {
    return location || '';
  }
};

/**
 * Generate an ICS for sending event updates.
 * @param event Event content.
 * @param status Status of the event.
 * @param method Calendar update type - REQUEST, REPLY, CANCEL.
 * @param fromAddress Sender of the update, used for matching attendees in reply below.
 * @returns Generated ICS.
 */
export const generateICS = (
  event: GenerateEvent,
  status: ICalEventStatus,
  method: ICalCalendarMethod,
  fromAddress: string
) => {
  const organizer = event.attendees.find((attendee) => attendee.permission === AttendeePermission.Owner);

  const calendar = ical({
    prodId: {
      company: ICAL_COMPANY,
      product: ICAL_PRODUCT,
      language: ICAL_LANGUAGE
    },
    scale: 'GREGORIAN',
    method
  });
  const isReply = method === ICalCalendarMethod.REPLY;
  const ev = calendar.createEvent({
    start: new Date(event.startDate),
    end: new Date(event.endDate),
    timezone: getLocalTimezone(),
    stamp: new Date(),
    allDay: event.isAllDay,
    // Don't send the organizer if it's a reply.
    organizer:
      organizer && !isReply
        ? {
            name: organizer.displayName || organizer.email,
            email: organizer.email
          }
        : undefined,
    summary: event.title,
    description: event.description,
    location: constructLocationString(event.location, event.conference),
    status,
    id: event.externalID,
    // TODO(easdar) pipe createdAt through.
    created: new Date(),
    lastModified: new Date(event.updatedAt),
    sequence: event.sequence,
    transparency: ICalEventTransparency.OPAQUE,
    x: event.conference && isURL(event.conference) ? [{ key: CONFERENCE_KEY, value: event.conference }] : [],
    repeating: event.recurrenceRule ? event.recurrenceRule.toICalGenerator() : undefined,
    recurrenceId: event.recurrenceID
  });

  // For replies, we need to filter out all attendees execept for the one we are updating the status of.
  event.attendees
    .filter((attendee) => {
      if (isReply && !(attendee.email === fromAddress)) {
        return false;
      }
      return true;
    })
    .map((attendee) => {
      ev.createAttendee({
        email: attendee.email,
        name: attendee.displayName,
        role: ICalAttendeeRole.REQ,
        rsvp: true,
        status: transformStatus(attendee.attendeeStatus)
      });
    });

  // We only generate ics for events when sent to the event's participants, so in this case we include our default reminders
  // Notification at 1 day before at 9 am if all day event, in seconds
  const oneDayBeforeAt9AMTrigger = DEFAULT_ALL_DAY_REMINDER_HOURS_BEFORE * MINUTES_IN_HOUR * SECONDS_IN_MIN;
  // Notification at 10 minutes before the event if timed event, in seconds
  const tenMinutesBeforeTrigger = DEFAULT_TIMED_REMINDER_MINS_BEFORE * SECONDS_IN_MIN;
  const triggerValue = event.isAllDay ? oneDayBeforeAt9AMTrigger : tenMinutesBeforeTrigger;

  ev.createAlarm({
    type: ICalAlarmType.display, // For a notification
    trigger: triggerValue
  });

  return calendar.toString();
};
