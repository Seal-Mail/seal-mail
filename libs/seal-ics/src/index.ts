export { parseICS, ParsedEvent } from './parse';
export { RecurrenceRule, MAX_RECURRENCE_COUNT } from './RecurrenceRule';
export { generateICS, GenerateAttendee, GenerateEvent } from './generate';
export * from './types';
export * from './utils';
