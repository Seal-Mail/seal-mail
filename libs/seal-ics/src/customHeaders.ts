import { ParamList } from 'ical';
import isURL from 'validator/lib/isURL';

import { CONFERENCE_KEY } from './constants';

const stripKey = (key: string) => key.replace('X-', '');

export const extractConference = (event: { [prop: string]: string | ParamList | undefined }) => {
  const conference = event[stripKey(CONFERENCE_KEY)];
  if (conference && typeof conference === 'string' && isURL(conference)) return conference;
};
