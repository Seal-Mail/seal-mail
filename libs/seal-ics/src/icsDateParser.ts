import dayjs from 'dayjs';
import timezone from 'dayjs/plugin/timezone';
import utc from 'dayjs/plugin/utc';
import { findIana } from 'windows-iana';

import { GMT_TZID_PATTERN } from './constants';
import { ExtendedDate } from './types';

/**
 * ical.js is missing timezone logic when parsing .ics files.
 *
 * This file overrides the start+end date parsing functions of icl.js,
 * it is instead adds a parsing function very similar to the origin but with timezones parsing.
 *
 * This ical bug appears in several issues and have few possible solutions suggested,
 * this one is the most lean and strait forward solution,
 * It only uses dayjs timezone parsing, while avoiding holding/getting timezones lists and handling them manually.
 */

dayjs.extend(utc);
dayjs.extend(timezone);

type Value = string | ExtendedDate | (string | ExtendedDate)[];

const text = function (t: string): string {
  t = t || '';
  return t
    .replace(/\\\,/g, ',')
    .replace(/\\\;/g, ';')
    .replace(/\\[nN]/g, '\n')
    .replace(/\\\\/g, '\\');
};

const parseValue = function (val: string) {
  if ('TRUE' === val) return true;

  if ('FALSE' === val) return false;

  const number = Number(val);
  if (!isNaN(number)) return number;

  return val;
};

const parseParams = function (p: string) {
  const out: { [key: string]: string | number | boolean } = {};
  for (let i = 0; i < p.length; i++) {
    const param = p[i];
    if (param && param.indexOf('=') > -1) {
      const segs = param.split('=');
      const key = segs[0] || '';
      out[key] = parseValue(segs.slice(1).join('='));
    }
  }
  return out;
};

const storeValParam = function (name: string) {
  return function (val: string | ExtendedDate, curr: { [key: string]: Value }) {
    const current = curr[name];
    if (Array.isArray(current)) {
      current.push(val);
      return curr;
    }

    if (current != null) {
      curr[name] = [current, val];
      return curr;
    }

    curr[name] = val;
    return curr;
  };
};

/** Check if timezone is valid IANA timezone and try converting from Windows time zone to IANA time zone if not */
const parseTZ = (inputTZID: string) => {
  // Remove surrounding quotes if found at the beginning and at the end of the string
  // (Occurs when parsing Microsoft Exchange events containing TZID with Windows standard format instead IANA)
  let parsedTZID: string | undefined = inputTZID.replace(/^"(.*)"$/, '$1');
  let utcOffset = 0;

  // Microsoft sometimes uses this timezone which is not iana timezone
  // https://www.jevents.net/discussions/datetimezone-construct-unknown-or-bad-timezone-tzone-microsoft-utc
  if (parsedTZID === 'tzone://Microsoft/Utc') {
    parsedTZID = 'UTC';
  }

  try {
    new Date().toLocaleString('en-US', { timeZone: parsedTZID });
  } catch (e) {
    // Convert to IANA time zone
    const results = findIana(parsedTZID);
    if (results && results.length > 0 && results[0]) {
      parsedTZID = results[0];
    } else {
      const errorMsg = `Invalid timezone: ${parsedTZID}`;

      // If not IANA or Windows timezone, check for GMT+0100 format
      const regResult = RegExp(GMT_TZID_PATTERN).exec(parsedTZID);
      const gmtTZID = regResult?.[0];
      if (!gmtTZID) throw new Error(errorMsg); // if not defined, the TZID is not in GMT+0100 format

      // Extract utc offset from the TZID
      const offsetString = gmtTZID.replace(/GMT/, '');
      // the offset will only be the first 3 chars, ie +01 from +0100
      utcOffset = parseInt(offsetString.slice(0, 3));
      if (isNaN(utcOffset)) throw new Error(errorMsg);
      parsedTZID = undefined;
    }
  }

  return { parsedTZID, utcOffset };
};

const addTZ = function (dt: ExtendedDate, params: string) {
  const p = parseParams(params);

  if (params && p) {
    dt.tz = p.TZID;
    if (dt.tz !== undefined) {
      // Check if time zone is in Windows format rather than IANA format and parse to IANA if necessary
      const parsedTZID = parseTZ(dt.tz as string).parsedTZID;
      dt.tz = parsedTZID;
    }
  }

  return dt;
};

export const dateParam = function (name: string) {
  return function (val: string, params: string, curr: { [key: string]: Value }) {
    let newDate: string | ExtendedDate = text(val);
    if (params && params.indexOf('VALUE=DATE') > -1 && params.indexOf('VALUE=DATE-TIME') === -1) {
      // Just Date
      // console.log(" date string="+val)
      const comps = /^(\d{4})(\d{2})(\d{2}).*$/.exec(val);
      if (comps !== null) {
        // No TZ info - assume same timezone as this computer
        newDate = new Date(parseInt(comps[1]!, 10), parseInt(comps[2]!, 10) - 1, parseInt(comps[3]!, 10));

        newDate = addTZ(newDate, params);
        newDate.dateOnly = true;

        // Store as string - worst case scenario
        return storeValParam(name)(newDate, curr);
      }
    }

    // Typical RFC date-time format
    const comps = /^(\d{4})(\d{2})(\d{2})T(\d{2})(\d{2})(\d{2})(Z)?$/.exec(val);
    if (comps !== null) {
      if (comps[7] == 'Z') {
        // GMT
        newDate = new Date(
          Date.UTC(
            parseInt(comps[1]!, 10),
            parseInt(comps[2]!, 10) - 1,
            parseInt(comps[3]!, 10),
            parseInt(comps[4]!, 10),
            parseInt(comps[5]!, 10),
            parseInt(comps[6]!, 10)
          )
        );
      } else {
        const p = parseParams(params);
        ///////////////////////////////////////////////////////////////////////////////////////////////////
        // This block is the only part changing in this patch
        if (p && p.TZID) {
          // If timezone is specified, the value corresponds to time local to the timezone.
          // So, timezone needs to be factored in before converting to UTC
          // Ref: #3 option in https://www.kanzaki.com/docs/ical/dateTime.html
          const paramTZID: string | undefined = p.TZID as string;
          const { parsedTZID, utcOffset } = parseTZ(paramTZID);

          const newDayjsWithTimezone = utcOffset ? dayjs(val).utcOffset(utcOffset, true) : dayjs.tz(val, parsedTZID);
          newDate = newDayjsWithTimezone.utc().toDate();

          ////////////////////////////////////////////////////////////////////////////////////////////////////
        } else {
          newDate = new Date(
            parseInt(comps[1]!, 10),
            parseInt(comps[2]!, 10) - 1,
            parseInt(comps[3]!, 10),
            parseInt(comps[4]!, 10),
            parseInt(comps[5]!, 10),
            parseInt(comps[6]!, 10)
          );
        }
      }

      newDate = addTZ(newDate, params);
    }

    // Store as string - worst case scenario
    return storeValParam(name)(newDate, curr);
  };
};
