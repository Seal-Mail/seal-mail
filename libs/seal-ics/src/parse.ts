import dayjs, { Dayjs } from 'dayjs';
import ical, { ParamList } from 'ical';
import { EventReminderType, EventReminderTimeUnit, EventReminder } from 'skiff-graphql';
import { v4 } from 'uuid';

import { METHOD_REGEX_PATTERN } from './constants';
import { extractConference } from './customHeaders';
import { dateParam } from './icsDateParser';
import { RecurrenceRule } from './RecurrenceRule';
import {
  ParsedOrganizer,
  Params,
  ParsedAttendee,
  CalendarMethodTypes,
  AttendeeStatParser,
  CalendarUserTypesParser,
  EventStatus,
  ExtendedCalendarComponent
} from './types';
import { convertReminderToUTC, isAllDay } from './utils';

/**
 * override ical date parsing with custom parsing
 */
// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
(ical as any).objectHandlers.DTSTART = dateParam('start');
// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
(ical as any).objectHandlers.DTEND = dateParam('end');
// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
(ical as any).objectHandlers['RECURRENCE-ID'] = dateParam('recurrenceid');

type ParseError = {
  id: string;
  component: ExtendedCalendarComponent;
  error: string;
};

export interface ParsedEvent {
  id: string;
  startDate: Dayjs;
  endDate: Dayjs;
  icsCreationDate?: Dayjs;
  creationDate?: Dayjs;
  lastModifiedDate?: Dayjs;
  organizer?: ParsedOrganizer;
  attendees: ParsedAttendee[];
  description: string;
  location: string;
  title: string;
  status: EventStatus;
  method: CalendarMethodTypes;
  sequence: number;
  reminders?: EventReminder[];
  // recurrence
  recurrenceRule?: RecurrenceRule;
  recurrenceId?: Dayjs;
  recurrences?: ParsedEvent[];
  conference?: string;
}

interface ParsedICS {
  events: ParsedEvent[];
  errors: ParseError[];
  method: CalendarMethodTypes;
}

interface VAlarm {
  type: 'VALARM';
  action: string;
  trigger: string; // ISO8601 duration or explicit datetime
  attendee?: string;
  description?: string;
  summary?: string;
}
const DURATION_REGEX = /-?P(?:(\d+)W)?(?:(\d+)D)?T?(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?/;

/*
ical's event.start and event.end has a dateOnly attr but types of ical
are not correct (type is Date).
*/
type ICalEventDate = Date & { dateOnly: boolean };

export const removeQuotes = (str: string) => str.replace(/^\"(.*)\"$/, '$1');

const parseOrganizer = (data: ParamList | undefined | string): ParsedOrganizer | undefined => {
  if (!data || typeof data === 'string') return undefined;

  return {
    name: removeQuotes(data.params[Params.Name] || ''),
    language: data.params[Params.Language],
    sentBy: data.params[Params.SentBy],
    value: data.val
  };
};

const parseAttendees = (data: ParamList | undefined | string | ParamList[]): ParsedAttendee[] => {
  if (!data || typeof data !== 'object') return [];

  // sometime single attendee received as single ParamList object
  if (!Array.isArray(data)) data = [data];

  const attendees: ParsedAttendee[] = [];

  data.forEach((attendeeData) => {
    attendees.push({
      name: removeQuotes(attendeeData.params[Params.Name] || ''),
      value: attendeeData.val,
      email: attendeeData.val.replace(new RegExp('mailto:', 'ig'), ''),
      sentBy: attendeeData.params[Params.SentBy],
      language: attendeeData.params[Params.Language],
      role: attendeeData.params[Params.Role] || '',
      status: AttendeeStatParser.parse(attendeeData.params[Params.ParticipantStatue]?.toUpperCase()),
      rsvp: attendeeData.params[Params.Rsvp] === 'TRUE' ? true : false,
      calendarUserType: CalendarUserTypesParser.parse(attendeeData.params[Params.CalendarUserType]),
      memberships: attendeeData.params[Params.Member],
      delegatedTo: attendeeData.params[Params.DelegatedTo],
      delegatedFrom: attendeeData.params[Params.DelegatedFrom]
    });
  });
  return attendees;
};

const parseToDayJS = (value: Date | undefined, name: string): Dayjs => {
  if (!value) throw new Error(`Event has no ${name} time`);
  try {
    return dayjs(value);
  } catch (e) {
    console.error(e);
    throw new Error(`Failed parsing ${name} time`);
  }
};

const parseMethodParam = (ics: string) => {
  const regResult = RegExp(METHOD_REGEX_PATTERN).exec(ics);
  // if method is not defined, default to Publish
  if (regResult === null) return CalendarMethodTypes.Publish;
  // if found method string in ics, get the method value
  if (regResult?.[0]) return regResult[0].split(':')[1] as CalendarMethodTypes;
  else throw new Error('invalid ICS method was found on ICS file');
};

const parseGeneric = (value: ParamList | undefined | string, defaultValue = ''): string => {
  switch (typeof value) {
    case 'object':
      return value.val;
    case 'string':
      return value;
    default:
      return defaultValue;
  }
};

const getSecondsFromDurationStr = (durationFromICS: string) => {
  try {
    let seconds = 0;
    const durationStr = durationFromICS.toLowerCase();
    const days = durationStr.match(/(\d+)\s*d/);
    const hours = durationStr.match(/(\d+)\s*h/);
    const minutes = durationStr.match(/(\d+)\s*m/);
    if (days && days[1]) {
      seconds += parseInt(days[1]) * 86400;
    }
    if (hours && hours[1]) {
      seconds += parseInt(hours[1]) * 3600;
    }
    if (minutes && minutes[1]) {
      seconds += parseInt(minutes[1]) * 60;
    }
    return seconds;
  } catch (error) {
    // return 1 hour in seconds
    return 3600;
  }
};

const parseReminder = (alarm: VAlarm, isAllDayEvent: boolean, startDate: Dayjs, userTimezone?: string) => {
  const mediumFullHourFormat = 'H:mm';
  const trigger = alarm.trigger;
  let timeValue: number, timeUnit: EventReminderTimeUnit;

  // Match the duration components
  const match = DURATION_REGEX.exec(trigger);
  if (!match) return undefined;

  // Extract the duration components
  const weeks = (match[1] && parseInt(match[1], 10)) || 0;
  const days = (match[2] && parseInt(match[2], 10)) || 0;
  const hours = (match[3] && parseInt(match[3], 10)) || 0;
  const minutes = (match[4] && parseInt(match[4], 10)) || 0;
  const seconds = match[5] && parseInt(match[5], 10);

  // If seconds equal zero, this means the notification is set at start of the event
  if (!minutes && !hours && !days && !weeks && seconds !== 0) return undefined;

  const type = alarm.action === EventReminderType.Email ? EventReminderType.Email : EventReminderType.Notification;

  timeValue = weeks || days || hours || minutes || 0;
  // Determine the time unit and value based on the duration components
  if (weeks) {
    timeUnit = EventReminderTimeUnit.Week;
  } else if (days) {
    timeUnit = EventReminderTimeUnit.Day;
  } else if (hours) {
    timeUnit = EventReminderTimeUnit.Hour;
  } else if (minutes || seconds === 0) {
    timeUnit = EventReminderTimeUnit.Minute;
  } else {
    return undefined;
  }

  let reminder: EventReminder = {
    timeUnit,
    type,
    timeValue,
    timeForAllDay: undefined,
    reminderID: v4()
  };

  if (!isAllDayEvent) return reminder;

  // In all day events we need to get the time from the ISO 8601 duration string
  // Time is the event's start date minus the hours in trigger
  const timeForAllDay = startDate.subtract(hours, 'hour').format(mediumFullHourFormat);

  if (days && hours) {
    // If there are days and hours, we add 1 to the number of days to account for those extra hours
    // For example P2DT16H0M0S means 2 days and 16 hours before event which we should save as 3 days before at H:mm
    timeValue = timeValue + 1;
    // Some calendars eg: google and apple doesn't include weeks in ics files, so instead we can have 21 days before event
    // We check how many whole weeks in the given number of days if any
    if (days % 7 === 0) {
      timeUnit = EventReminderTimeUnit.Week;
      timeValue = days / 7;
    }
  } else if (hours) {
    // If only hours are set in the trigger eg: 15 hrs before event, we save it as 1 day before at 9am
    timeUnit = EventReminderTimeUnit.Day;
    timeValue = 1;
  }

  reminder = {
    ...reminder,
    timeUnit,
    timeValue,
    timeForAllDay
  };

  return userTimezone ? convertReminderToUTC(reminder, userTimezone) : reminder;
};

const parseEvent = (
  id: string,
  event: ExtendedCalendarComponent,
  method: CalendarMethodTypes,
  userTimezone?: string
): ParsedEvent => {
  const recurrences: ParsedEvent[] = [];
  const recurrenceID = event.recurrenceid ?? event._recurrenceid;
  for (const recurrenceDate in event.recurrences) {
    const recurrenceEvent: ExtendedCalendarComponent | undefined =
      event.recurrences[recurrenceDate as unknown as number];
    // Skip if recurrence event is the same as the original event
    // This is a bug in ical.js
    if (!recurrenceEvent || recurrenceEvent.recurrenceid === recurrenceID) continue;
    recurrences.push(parseEvent(recurrenceEvent.uid || '', recurrenceEvent, method, userTimezone));
  }

  const excludeDates: Date[] = [];
  for (const exDate in event.exdate) {
    const excludeDate = event.exdate[exDate as unknown as number];
    if (!excludeDate) continue;
    excludeDates.push(excludeDate);
  }

  const startDate = parseToDayJS(event.start, 'start');
  let endDate: Dayjs | undefined = undefined;

  // for edge-cases where event.start is dateOnly but not event.end, so it's an all-day-event (single day)
  if (event.end && (event.start as ICalEventDate).dateOnly && !(event.end as ICalEventDate).dateOnly) delete event.end;

  if (event.end) {
    endDate = parseToDayJS(event.end, 'end');
  } else if (event.duration && typeof event.duration === 'string') {
    const durationSeconds = getSecondsFromDurationStr(event.duration);
    endDate = startDate.add(durationSeconds, 'second');
  } else if ((event.start as ICalEventDate).dateOnly) {
    // fallback end for end date (if start is dateOnly) - 1 day after start date
    endDate = startDate.add(1, 'day');
  } else {
    // fallback end for end date - 1 hour after start date
    endDate = startDate.add(1, 'hour');
  }

  let sequence = 0;
  try {
    sequence = event.sequence ? Number(event.sequence as string) : 0;
  } catch {
    sequence = 0;
  }
  const isAllDayEvent = isAllDay(startDate, endDate);
  const recurrenceRule = event.rrule ? RecurrenceRule.fromRRule(event.rrule, excludeDates, isAllDayEvent) : undefined;
  if (recurrenceRule && !recurrenceID) {
    recurrenceRule.startDate = startDate.valueOf();
  }

  const reminders: EventReminder[] = [];
  let parsedReminder: EventReminder | undefined;

  for (const key in event) {
    const component = event[key];
    // Check if an object is a reminder
    if (component && typeof component === 'object' && 'type' in component && component?.type === 'VALARM') {
      // Get parsed reminder and add it to our array of reminders
      parsedReminder = parseReminder(component as unknown as VAlarm, isAllDayEvent, startDate, userTimezone);
      if (!parsedReminder) continue;

      reminders.push(parsedReminder);
    }
  }

  return {
    id,
    startDate,
    endDate,
    icsCreationDate: event.dtstamp && parseToDayJS(event.dtstamp, 'dtstamp'),
    creationDate: event.created && parseToDayJS(event.created, 'created'),
    lastModifiedDate: event.lastmodified && parseToDayJS(event.lastmodified, 'lasymodified'),
    organizer: parseOrganizer(event.organizer),
    attendees: parseAttendees(event.attendee),
    description: parseGeneric(event.description),
    location: parseGeneric(event.location),
    title: parseGeneric(event.summary, 'Untitled event'),
    status: event.status as EventStatus,
    recurrenceRule,
    recurrenceId: recurrenceID ? parseToDayJS(recurrenceID as Date, 'recurrenceid') : undefined,
    recurrences,
    method,
    conference: extractConference(event),
    sequence,
    reminders
  };
};

export const parseICS = (ics: string, userTimezone?: string): ParsedICS => {
  const parsedICS = ical.parseICS(ics);
  const method = parseMethodParam(ics);
  const events: ParsedEvent[] = [];
  const errors: ParseError[] = [];

  for (const id in parsedICS) {
    const calendarComponent: ExtendedCalendarComponent | undefined = parsedICS[id];
    if (!calendarComponent) continue;
    if (calendarComponent.type !== 'VEVENT') continue;

    if (calendarComponent.rrule) {
      // We need to pass the event TZID to the recurrence rule
      // Otherwise, the recurring event TZID will default to 'UTC' which messes up event timing across DST periods
      calendarComponent.rrule.origOptions.tzid = calendarComponent.start?.tz as string | undefined;
    }

    try {
      const parsedEvent = parseEvent(id, calendarComponent, method, userTimezone);
      events.push(parsedEvent);
    } catch (err) {
      console.error(`Failed to parse event, ${(err as { message: string }).message}`, calendarComponent);
      errors.push({ id, component: calendarComponent, error: (err as { message: string }).message });
    }
  }

  return {
    events,
    errors,
    method
  };
};
