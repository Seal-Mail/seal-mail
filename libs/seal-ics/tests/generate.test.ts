import { ICalCalendarMethod, ICalEventStatus } from 'ical-generator';
import { AttendeePermission, AttendeeStatus } from 'skiff-graphql';

import { generateICS } from '../src/generate';

describe('test generate ICS', () => {
  jest.useFakeTimers().setSystemTime(new Date('2021-01-01'));
  it('should generate ICS', () => {
    const ics = generateICS(
      {
        title: 'test',
        description: 'test',
        location: 'test',
        sequence: 0,
        startDate: new Date('2021-01-01').getTime(),
        endDate: new Date('2021-01-02').getTime(),
        externalID: 'abcd',
        attendees: [
          {
            attendeeStatus: AttendeeStatus.Yes,
            permission: AttendeePermission.Owner,
            displayName: 'foo',
            optional: false,
            email: 'foo@skiff.town'
          },
          {
            attendeeStatus: AttendeeStatus.Yes,
            permission: AttendeePermission.Owner,
            displayName: 'bar',
            optional: false,
            email: 'bar@skiff.org'
          }
        ],
        updatedAt: new Date('2021-01-01').getTime()
      },
      ICalEventStatus.CONFIRMED,
      ICalCalendarMethod.REQUEST,
      'foo@skiff.town'
    );
    expect(ics).toMatchSnapshot();
  });
});
