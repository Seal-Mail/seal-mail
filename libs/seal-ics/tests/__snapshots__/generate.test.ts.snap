// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generate ICS should generate ICS 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Skiff//Skiff Calendar//EN
CALSCALE:GREGORIAN
METHOD:REQUEST
BEGIN:VEVENT
UID:abcd
SEQUENCE:0
DTSTAMP:20210101T000000Z
DTSTART:20210101T000000Z
DTEND:20210102T000000Z
SUMMARY:test
TRANSP:OPAQUE
LOCATION:test
DESCRIPTION:test
ORGANIZER;CN="foo":mailto:foo@skiff.town
ATTENDEE;ROLE=REQ-PARTICIPANT;PARTSTAT=ACCEPTED;RSVP=TRUE;CN="foo":MAILTO:
 foo@skiff.town
ATTENDEE;ROLE=REQ-PARTICIPANT;PARTSTAT=ACCEPTED;RSVP=TRUE;CN="bar":MAILTO:
 bar@skiff.org
BEGIN:VALARM
ACTION:DISPLAY
TRIGGER:-PT10M
DESCRIPTION:test
END:VALARM
STATUS:CONFIRMED
CREATED:20210101T000000Z
LAST-MODIFIED:20210101T000000Z
END:VEVENT
END:VCALENDAR"
`;
